Character creation form

McGarvelcuddy opened the pull request:
### Working in ```public\characterCreator.html``` to create basic form for character creation.

- [x] Create fields for all sections of form, initialized at 0
- [ ] Stats limited based on character level
- [x] Automatic calculations of skill point costs in JavaScript
- [ ] Automatic calculations of Health Points, Physical Defense, and Mental Defense
- [x] Cannot submit form if remaining stat points are negative
- [x] Stat fields have increment/decrement arrow controls inside
- [ ] CSS styling of page
- [ ] Eventual cookie detection

McGarvelcuddy committed:
Starting the form off

Added boilerplate html and basic form to ```public\characterCreator.html```

McGarvelcuddy committed:
Added fields to all sections

Added fields to all these boys

McGarvelcuddy committed:
Added classes to stats

Stat class added

HistoryWill committed:
Test commit?

HistoryWill committed:
Basic stat calculator, took xp out of stats class.

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,96 @@
\ No newline at end of file
+an interactive character creator
+<!DOCTYPE html>
+<!--[if lt IE 7]>      <html class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]-->
```
if any type of ie => just give a message saying to upgrade to a better browser

McGarvelcuddy replied:
Yea I need to clean it up, I basically left in all of the boilerplate stuff

McGarvelcuddy replied:
Using this now

`<!--[if IE]>
            <p class="browsehappy">You are using an <strong>outdated</strong> browser. Please <a href="#">upgrade your browser</a> to improve your experience.</p>
        <![endif]-->`

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,96 @@
+<!--[if gt IE 8]>      <html class="no-js"> <!--<![endif]-->
+<html>
+
+<head>
```
all of the head should be covered by the nav

McGarvelcuddy replied:
You'll have to show us how this is supposed to go

Anolan19 replied:
1. isn't Zack doing that bit, and 2. most likely through a mix of javascript and html imports

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,96 @@
+    </style>
+
+    <title></title>
+    <meta name="description" content="">
```
meta and style from the nav too

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,96 @@
+        <![endif]-->
+    <form id="characterCreator" name="characterCreator" action="../creator.js" method="POST">
+        <label>Name</label>
+        <input type="text" id="name" name="name"><br><br>
```
ids are for javascript and class names are for adding style, you should add descriptive class name(s) to them all

McGarvelcuddy replied:
So instead, I'll make ```id="stat"``` and then ```class="<something we define later when doing the styling>"```.  And we only need to keep their identifying marks in ```name``` for the submission of the form?

McGarvelcuddy replied:
Actually, do I even need id?

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,96 @@
+      console.log(elements);
+      let y = calculateDeficit(elements);
+
+      document.getElementById('deficitpoints').innerHTML = y;
```
document.getElementById('deficitpoints').innerHTML  = calculateDeficit(elements)

HistoryWill replied:
Replaced with your suggestion

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,96 @@
+      document.getElementById('deficitpoints').innerHTML = y;
+
+    }
+    function generateStatCost(x){
```
you can do this without let

Anolan19 replied:
i.e. in a way where you don't have to declare variables

HistoryWill replied:
Maybe did it correctly?

Anolan19 replied:
Huh?

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,96 @@
+    function calculateDeficit(stat){
+      let y = 0;
+      for(x of stat){
+        console.log(x.value);
```
when you merge, no console statements

HistoryWill replied:
Removed the console statements

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,96 @@
+      }
+      return res;
+    }
+    function calculateDeficit(stat){
```
what is the point of this? the bonus points are equal to the total negative

HistoryWill replied:
Can you elaborate further? I feel like I'm making a fundemantal misunderstanding.

Anolan19 replied:
If only the negative values give points, why not include that calculation in the point cost calculation, and have that return a negative cost, i.e. a refund?

McGarvelcuddy replied:
oh wow, looking at the solution in there now, I never thought you could do multiple declaration/initialization of variables within the declaration of a for-loop

HistoryWill committed:
Making corrections and removing console statements.

HistoryWill committed:
Edited function statcost, fixed var names in both funcs relate to stats

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,107 @@
+    }
+
+    </script>
+    <h3>Statpoints</h3>
```
https://www.w3.org/MarkUp/html3/headings.html#:~:text=HTML%20defines%20six%20levels%20of,level%20and%20H6%20the%20least.

Anolan19 replied:
I meant for this to be a link to a wcag doc, but it didn't link to that page, you should look up aria/ accessibility for semantic meaning for elements

McGarvelcuddy committed:
Should make a cookie with a JSON string inside

When submitting the form, it should instead make a cookie with the information from the form all saved in a stringified JSON object

McGarvelcuddy committed:
Added es-dev-server for easy dev environment testing

Initialized package.json and defined "npm start" script.  Going to pull this into character sheet

McGarvelcuddy committed:
Merge pull request #2 from Anolan19/dev_environment

Added es-dev-server for easy dev environment testing

McGarvelcuddy committed:
Skeleton of character sheet

McGarvelcuddy committed:
Creating a character directs to the character sheet

McGarvelcuddy committed:
Creation Restriction and Documentation

Altered InternetExplorer detection
Added condition to prevent creation when points will be negative
Remaining points update in real time while user changes form
Redirects to characterSheet.html page upon character creation
Documented my functions

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,157 @@
+        <![endif]-->
+    <form id="characterCreator">
+        <label>Name</label>
+        <input type="text" id="name" name="name" class=""><br><br>
```
class=""?

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,157 @@
+            <p class="browsehappy">You are using an <strong>outdated</strong> browser. Please <a href="#">upgrade your browser</a> to improve your experience.</p>
+        <![endif]-->
+    <form id="characterCreator">
+        <label>Name</label>
```
label for?

McGarvelcuddy replied:
Done

Anolan19 commented on public/characterSheet.html:
```diff
@@ -1 +1,47 @@
+    <script></script>
+</body>
+
+</html>
```
no trailing newline?

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,157 @@
+            if (setStatLevel < 0) { //Checks if setstatlevel is negative if so returns the negative setstatlevel as a statpoint gain
+                return setStatLevel
+            } else {
+                for (loopIterator = 0, result = 0; loopIterator <= setStatLevel; loopIterator++) { // iterates to find the statpoint cost needed to get a stat to the setstat level
```
maybe look into this, you don't need to use it
https://scotch.io/tutorials/understanding-memoization-in-javascript#toc-a-functional-approach

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,157 @@
+        function calculateDeficit(statCollection) {
+            let workingValue = 0;
+            for (statValue of statCollection) {
+                if (isNaN(statValue.value)) {
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,157 @@
+                setCookie("value", value, 365);
+
+                //and take us to the character sheet page to see our new character stats
+                window.location.href = "characterSheet.html";
```
Have you considered instead of doing it like this, having it be a submit handler on the form and not setting the href? I think that doing the submit handler would be better code style/practice https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/submit_event

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,157 @@
+         * @param {integer} exdays Number of days the browser should store this cookie.  Broken down from days into milliseconds
+         */
+        function setCookie(cname, cvalue, exdays) {
+            var d = new Date();
```
var

Anolan19 replied:
d

Anolan19 replied:
no comments?

McGarvelcuddy replied:
made it all inline, and commented on the static multiplication

Anolan19 commented on public/characterCreator.html:
```diff
@@ -1 +1,157 @@
+    </script>
+</body>
+
+</html>
```
no trailing newline?

McGarvelcuddy committed:
Character sheet now displays character info from cookie

Anolan19 commented:
https://stackoverflow.com/questions/32598971/whats-the-purpose-of-if-typeof-window-undefined

McGarvelcuddy committed:
Removed var, made things inline, and documented

McGarvelcuddy committed:
Added for section to labels in the form
